### ISSUE TITLE
`balena deploy` should parse, validate, and create a contract as part of the release it creates

### ISSUE THREAD

**Issue by sradevski** [balena-io/balena-cli#2183]
It should essentially do the same as here: https://github.com/balena-io/balena-builder/blob/master/src/app-contract.ts

**Comment by pdcastro** [comment]
There is a related product call topic pending discussion:

> https://jel.ly.fish/view-all-brainstorm-topics/brainstorm-topic-7478d22a-59b8-44a3-a5bf-eaaa247cb396
> 
> Discuss what to do (if anything) about the fact that two `balena.yml` files may exist in a balena app with different purposes and formats.
> ```
> ./balena.yml - contracts / Deploy With Balena
> ./.balena/balena.yml - build secrets and variables
> ```
> Example: the ARG and ENV instructions in a Dockerfile are used for build time variables and runtime environment variables respectively. To set build time variables, one should use `./.balena/balena.yml`, but to set environment variables, one should use `./balena.yml`. This separation seems difficult to justify, especially when both files have the same name. Also:
> * `./.balena/balena.yml` allows specifying service-specific variables, whereas `./balena.yml` only allows "global" (all services) variables, which is an inconsistency.
> * `./.balena/balena.yml` is supported across `balena push`, `balena build`, `balena deploy`, whereas `./balena.yml` is only supported by `balena push` (because it is only the cloud builder that interprets `./balena.yml`, not the CLI).
> 
> Files documented at:
> * https://www.balena.io/docs/learn/deploy/deploy-with-balena-button/
> * https://www.balena.io/docs/learn/deploy/deployment/#build-time-only-secret-file
>
> At least one user was confused by it and brought it to support / forums: https://forums.balena.io/t/git-push-and-buildargs/227059/13
> 
> Would we want to rename one of the files? Merge them? Leave it as it is?

**Comment by pdcastro** [comment]
I didn't investigate it yet, but we may need to implement support for local push as well - `balena push <ip-addr>`. For example, the contract file includes an `applicationEnvironmentVariables` section that should be honoured when pushing to a local device. I suspect this doesn't currently happen.

**Comment by sradevski** [comment]
@pdcastro yes it doesn't happen currently, but we might need support for it in the future, if the envvars are read (which would be really convenient). For now, `balena deploy` should suffice

**Comment by pdcastro** [comment]
Connects to JF Product Improvement: [Draft releases and release versioning](https://jel.ly.fish/8ea1c390-9a85-402d-978c-4d31dcb0d235) (restricted access)
